From 60d6b6ac126ce3b4b1c83e537f26f564900e2eda Mon Sep 17 00:00:00 2001
From: Sebastian Bazley <sebb@apache.org>
Date: Wed, 23 Mar 2011 12:50:15 +0000
Subject: [PATCH] COMPRESS-117 Certain tar files not recognised by
 ArchiveStreamFactory Check if first entry can be read as a Tar entry

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/compress/trunk@1084567 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                                  | 3 +++
 .../commons/compress/archivers/ArchiveStreamFactory.java | 9 +++++++++
 2 files changed, 12 insertions(+)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 2e30b20fa6c..cf4be7d8e55 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -45,6 +45,9 @@ The <action> type attribute can be add,update,fix,remove.
   </properties>
   <body>
     <release version="1.2" date="as in SVN" description="Release 1.2">
+      <action issue="COMPRESS-117" type="update" date="2011-03-23">
+        Certain tar files not recognised by ArchiveStreamFactory.
+      </action> 
       <action issue="COMPRESS-125" type="fix" date="2011-03-23">
         BZip2CompressorInputStream throws IOException if underlying stream returns available() == 0.
         Removed the check.
diff --git a/src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java b/src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java
index a3b93ccc3b3..048aeb94a56 100644
--- a/src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java
+++ b/src/main/java/org/apache/commons/compress/archivers/ArchiveStreamFactory.java
@@ -18,6 +18,7 @@
  */
 package org.apache.commons.compress.archivers;
 
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -216,6 +217,14 @@ public ArchiveInputStream createArchiveInputStream(final InputStream in)
             if (TarArchiveInputStream.matches(tarheader, signatureLength)) {
                 return new TarArchiveInputStream(in);
             }
+            // COMPRESS-117 - improve auto-recognition
+            try {
+                TarArchiveInputStream tais = new TarArchiveInputStream(new ByteArrayInputStream(tarheader));
+                tais.getNextEntry();
+                return new TarArchiveInputStream(in);
+            } catch (Exception e) { // can generate IllegalArgumentException as well as IOException
+                // ignored
+            }
         } catch (IOException e) {
             throw new ArchiveException("Could not use reset and mark operations.", e);
         }
